### ISSUE TITLE
Spread operator with string array in optional function parameter does not catch type mismatch

### ISSUE THREAD

**Issue by Woodz** [microsoft/TypeScript#56923]
### 🔎 Search Terms

"spread operator", "string array", "function", "parameter", "optional", "rest"

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about spread operator and rest parameters

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.4.0-dev.20240101#code/GYVwdgxgLglg9mABMOcDqMoAsCyBDMAEzyjgCcBPABTzLwFsBTKRsgHgBUA+AChjAAOIKAC5EASRZ0ARgBtGnLgEpEAbwC+AWABQoSLATJUGbAHkBBsHlk06TKYr6DhAfjGTWeOQu4qNOnQgEAGcoRDAQJjIYCABBMjoKRABeRABtAEYAGkQAJhyAZgBdAG5AkLDQ6LAAc3jElPSAchQ4Jpym6Vp2xE68AC8m0oDdY0xcAmJSSlsGZlYeCKiY+rwKJTLR9HH8IhJyalo5qR4q-jqEtY2dAHobxABlLDgQWUJELDwAN0ZEIPoBDB5GREKwyOQdK0TBM9tNDnZ5mQeAA6VFLVgrS7rTZ3R7PV7vT4-P5wAFA1ighIQrbQ3ZTA6zewLVHIs61VbYkZQ8bmSzWRmIxaRDFxLHXGk8izwKw2I5MpFsi6JcW4p4vN4fb6-f6A4GU8FkSFjMxShD8uWClno6Ki5U4+4AUQASk7TE6xGqCZriTrySCweRENJhIhCHBGMFwnAoEbtia+bKEScWYqORsgA

### 💻 Code

```ts
function fooWithMandatoryParameter<T>(input: Iterable<T>) {}
function fooWithOptionalParameter<T>(input?: Iterable<T>) {}

const numericArray = [1, 2, 3];
const stringArray = ['foo', 'bar', 'baz'];

fooWithMandatoryParameter(numericArray);
fooWithMandatoryParameter(stringArray);
// Should have compiler error
fooWithMandatoryParameter(...numericArray);
// Should have compiler error
fooWithMandatoryParameter(...stringArray);

fooWithOptionalParameter(numericArray);
fooWithOptionalParameter(stringArray);
// Should have compiler error
fooWithOptionalParameter(...numericArray);
// ERROR: Should have compiler error but does not
fooWithOptionalParameter(...stringArray);
```

### 🙁 Actual behavior

`fooWithOptionalParameter(...stringArray);` was not caught by TS compiler

### 🙂 Expected behavior

`fooWithOptionalParameter(...stringArray);` should raise a compiler error that function needs rest argument to match a spread array

**Comment by fatcerberus** [comment]
Strings are legal `Iterable<string>` so that's why there's no error here

**Comment by Woodz** [comment]
> Strings are legal `Iterable<string>` so that's why there's no error here

This would allow `fooWithOptionalParameter(...'foo')`, but we are spreading an array of strings, not a single string.

I thought that using the spread operator to pass an array into a function would only work if the function used `rest` parameters (see https://stackoverflow.com/a/34181899/323177). Otherwise we are spreading an array of unknown length into function that only takes a single parameter - which should not be legal unless we know the array has exactly one element (not the case in the reproduced code).

In addition, the fact that this only happens with an optional parameter does not seem logical and implies it is a bug, rather than as designed.

**Comment by fatcerberus** [comment]
> This would allow `fooWithOptionalParameter(...'foo')`, but we are spreading an array of strings, not a single string.

The code as written is equivalent to calling `fooWithOptionalParameter('foo', 'bar', 'baz')`.  You possibly passed extra arguments, but the first one is a valid `Iterable<string>`, so TS is okay with it.  The other possibility would be that the array is completely empty - in which case the call becomes `fooWithOptionalParameter()`, which is also legal.

> In addition, the fact that this only happens with an optional parameter does not seem logical and implies it is a bug, rather than as designed.

Well, the version with the mandatory parameter requires at least one value (of type `Iterable<T>`) but the array might be empty.  On the other hand the optional version is safe to call with no arguments, and if you happen to pass extra ones, they'll just be quietly ignored.  So I'd say this falls under the same rationale that justifies https://github.com/microsoft/TypeScript/wiki/FAQ#why-are-functions-with-fewer-parameters-assignable-to-functions-that-take-more-parameters

**Comment by Woodz** [comment]
> The code as written is equivalent to calling `fooWithOptionalParameter('foo', 'bar', 'baz')`. You possibly passed extra arguments, but the first one is a valid `Iterable<string>`, so TS is okay with it.

TS is not ok with the extra arguments:
![image](https://github.com/microsoft/TypeScript/assets/966157/309df069-67d3-48e8-a5af-9fc0ef621f0c)


> On the other hand the optional version is safe to call with no arguments, and if you happen to pass extra ones, they'll just be quietly ignored. So I'd say this falls under the same rationale that justifies https://github.com/microsoft/TypeScript/wiki/FAQ#why-are-functions-with-fewer-parameters-assignable-to-functions-that-take-more-parameters

Quietly ignoring parameter mismatch does not sound like a good idea and is not the case when calling normal functions. The link you shared is for a very specific scenario where we are looking at assignability between function types. I do not think this is applicable in this case - especially given that if the parameter is mandatory, the error message matches my expectations - spreading an array is only allowed if the function accepts a `rest` parameter:
![image](https://github.com/microsoft/TypeScript/assets/966157/8752f28b-8faa-4694-a42a-740a39e8b561)

**Comment by fatcerberus** [comment]
After a quick search, I found this very similar issue: https://github.com/microsoft/TypeScript/issues/42119.  In fact it's almost identical to the case here since `Set`'s constructor is defined as taking an optional `Iterable<T>` parameter.
In particular, this comment https://github.com/microsoft/TypeScript/issues/42119#issuecomment-754965078 seems relevant.

> As long as the call isn't **provably** too-high arity, it should be OK, since **this is generally inline without our behavior in other places.**

Emphasis mine. (that last bit is why I linked to the FAQ entry I did)

**Comment by RyanCavanaugh** [comment]
This is the intended behavior. As in the function indirection case, it's not a "guaranteed error" to pass too many parameters. Saliently, if this error did exist, it's even not clear how you would make it go away in the cases where you meant to do this.
